### PR TITLE
Move code that interacts with the framework from .cu to .cc

### PR DESCRIPTION
**Pull request by fwyzard**

### CODE REVIEW

**Comment by fwyzard** [comment]
This seems to fix the issue with the MessageLogger in CMSSW_11_0_0_pre13.

**Comment by fwyzard** [comment]
:construction: Validation running at fu-c2a02-35-02:/data/user/fwyzard/patatrack/validation/run.41ltTrJmEq ...

**Comment by VinInn** [comment]
I suspect this makes  "cudaCompat" trick not working as it relies on ```__CUDACC__```
I am surprised that compiles.
I would have preferred to keep the launch in a cu

**Comment by VinInn** [comment]
do not understand how it could work: who is compiling  the new .cc files? gcc?? 
maybe yes, but changing the kernels in pure C++ "thanks" to cudaCompat.
not sure WHAT cuda is "launching".
if it works it is absolute magic.  (or I have not understood what is going on)

**Comment by fwyzard** [comment]
Right,  in fact it doesn't work, I forgot to include the kernels in a .cu file...
```
cudaCheck(cudaLaunchKernel( (const void*)kernel, config.gridDim, config.blockDim, (void**)pointers, config.sharedMem, config.stream));
cudaErrorInvalidDeviceFunction: invalid device function
```

**Comment by fwyzard** [comment]
:construction: Validation running at fu-c2a02-35-02:/data/user/fwyzard/patatrack/validation/run.IxWZOftePH ...

File: RecoLocalTracker/SiPixelRecHits/plugins/PixelRecHits.cc [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+// CMSSW headers
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/cudaCheck.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/device_unique_ptr.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/launch.h"
+#include "RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelRawToClusterGPUKernel.h"
+#include "RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusteringConstants.h"
+
+#include "PixelRecHits.h"
+#include "gpuPixelRecHits.h"
```
Comment (VinInn) [inline RecoLocalTracker/SiPixelRecHits/plugins/PixelRecHits.cc]: still, do not understand how, w/o cudaCompat that should NOT supposed to be active in this case, gcc would swallow cuda syntax.
Do you have self-contained trivial unit test that demonstrates how to launch a kernel from plain c++  compiled with gcc?

File: RecoLocalTracker/SiPixelRecHits/plugins/PixelRecHits.cc [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    if (nHits) {
+      cudautils::launch(gpuPixelRecHits::setHitsLayerStart, {1, 256, 0, stream}, clusters_d.clusModuleStart(), cpeParams, hits_d.hitsLayerStart());
+
+      if (nHits >= TrackingRecHit2DSOAView::maxHits()) {
+        edm::LogWarning("PixelRecHitGPUKernel")
+            << "Hits Overflow " << nHits << " > " << TrackingRecHit2DSOAView::maxHits();
+      }
+
+      auto hws = cudautils::make_device_unique<uint8_t[]>(TrackingRecHit2DSOAView::Hist::wsSize(), stream);
+      cudautils::fillManyFromVector(
```
Comment (VinInn) [inline RecoLocalTracker/SiPixelRecHits/plugins/PixelRecHits.cc]: again: this is the cpu version here, I suspect (cudaCompat magic)

**Comment by VinInn** [comment]
strictly speaking the warning is obsolete
```
        // this cannot happen anymore
        if (h >= TrackingRecHit2DSOAView::maxHits())
          break;  // overflow...
```

as
```
 static constexpr uint32_t maxHits() { return gpuClustering::MaxNumClusters; }
```


but ok, we can detect that the number of clusters is == to the maximum (or even detector by detector because with the low thresholds is this that was happening) This can be done in
 SiPixelRecHitCUDA::produce

I can make a PR, but testing is difficult (ok can reduce maxhits and see if happens!)

**Comment by VinInn** [comment]
My proposal in #422

File: RecoLocalTracker/SiPixelRecHits/plugins/PixelRecHits.cc (hunk 2 above)
Comment (fwyzard) [reply RecoLocalTracker/SiPixelRecHits/plugins/PixelRecHits.cc]: right... I had not understood that the choice of CPU vs GPU is dictated only by the .cc vs .cu file extension.

Comment (VinInn) [reply RecoLocalTracker/SiPixelRecHits/plugins/PixelRecHits.cc]: At the moment it is the only "very simple' mechanism I found that is not intrusive (and I was working under the assumption that in any case a kernel shall be launched "by nvcc").
Of course one can try change the magic if we find a useful mechanism to launch kernels from code compiled by a standard c++ compiler.

**Comment by fwyzard** [comment]
Superseded by #422.

**Pull request closed without merging**